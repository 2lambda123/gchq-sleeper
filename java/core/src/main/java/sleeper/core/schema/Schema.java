/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.schema;

import sleeper.core.schema.type.PrimitiveType;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * A Schema describes the fields found in a particular table in a Sleeper instance.
 */
public class Schema {
    private final List<Field> rowKeyFields;
    private final List<Field> sortKeyFields;
    private final List<Field> valueFields;

    public Schema() {
        this.rowKeyFields = new ArrayList<>();
        this.sortKeyFields = new ArrayList<>();
        this.valueFields = new ArrayList<>();
    }

    private Schema(Builder builder) {
        rowKeyFields = new ArrayList<>(builder.validRowKeyFields());
        sortKeyFields = new ArrayList<>(builder.validSortKeyFields());
        valueFields = new ArrayList<>(builder.validValueFields());
        validateNoDuplicates(Stream.concat(rowKeyFields.stream(),
                Stream.concat(sortKeyFields.stream(), valueFields.stream())));
    }

    public static Builder builder() {
        return new Builder();
    }

    public void setRowKeyFields(List<Field> rowKeyFields) {
        validateRowKeys(rowKeyFields);
        this.rowKeyFields.clear();
        this.rowKeyFields.addAll(rowKeyFields);
    }

    public void setRowKeyFields(Field... rowKeyFields) {
        setRowKeyFields(Arrays.asList(rowKeyFields));
    }

    public List<Field> getRowKeyFields() {
        return rowKeyFields;
    }

    public List<PrimitiveType> getRowKeyTypes() {
        return rowKeyFields.stream().map(Field::getType).map(t -> (PrimitiveType) t).collect(Collectors.toList());
    }

    public Field getSingleRowKeyField() {
        if (rowKeyFields.size() != 1) {
            throw new IllegalStateException("Cannot get single row key field, have " + rowKeyFields.size());
        }
        return rowKeyFields.get(0);
    }

    public void setSortKeyFields(List<Field> sortKeyFields) {
        validateSortKeys(sortKeyFields);
        this.sortKeyFields.clear();
        this.sortKeyFields.addAll(sortKeyFields);
    }

    public void setSortKeyFields(Field... sortKeyFields) {
        setSortKeyFields(Arrays.asList(sortKeyFields));
    }

    public List<Field> getSortKeyFields() {
        return sortKeyFields;
    }

    public List<PrimitiveType> getSortKeyTypes() {
        return sortKeyFields.stream().map(Field::getType).map(t -> (PrimitiveType) t).collect(Collectors.toList());
    }

    public void setValueFields(List<Field> valueFields) {
        this.valueFields.clear();
        this.valueFields.addAll(valueFields);
    }

    public void setValueFields(Field... valueFields) {
        setValueFields(Arrays.asList(valueFields));
    }

    public List<Field> getValueFields() {
        return valueFields;
    }

    public List<String> getRowKeyFieldNames() {
        return rowKeyFields.stream().map(Field::getName).collect(Collectors.toList());
    }

    public List<String> getSortKeyFieldNames() {
        return sortKeyFields.stream().map(Field::getName).collect(Collectors.toList());
    }

    public List<String> getValueFieldNames() {
        return valueFields.stream().map(Field::getName).collect(Collectors.toList());
    }

    public List<String> getAllFieldNames() {
        List<String> allFieldNames = new ArrayList<>();
        allFieldNames.addAll(getRowKeyFieldNames());
        allFieldNames.addAll(getSortKeyFieldNames());
        allFieldNames.addAll(getValueFieldNames());
        return allFieldNames;
    }

    public List<Field> getAllFields() {
        List<Field> allFields = new ArrayList<>();
        allFields.addAll(rowKeyFields);
        allFields.addAll(sortKeyFields);
        allFields.addAll(valueFields);
        return allFields;
    }

    public Optional<Field> getField(String fieldName) {
        return getAllFields().stream()
                .filter(f -> f.getName().equals(fieldName))
                .findFirst();
    }

    @Override
    public String toString() {
        return "Schema{" + "rowKeyFields=" + rowKeyFields + ", sortKeyFields=" + sortKeyFields + ", valueFields=" + valueFields + '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Schema schema = (Schema) o;

        return Objects.equals(rowKeyFields, schema.rowKeyFields) &&
                Objects.equals(sortKeyFields, schema.sortKeyFields) &&
                Objects.equals(valueFields, schema.valueFields);
    }

    @Override
    public int hashCode() {
        return Objects.hash(rowKeyFields, sortKeyFields, valueFields);
    }

    public static final class Builder {
        private List<Field> rowKeyFields;
        private List<Field> sortKeyFields;
        private List<Field> valueFields;

        private Builder() {
        }

        public Builder rowKeyFields(List<Field> rowKeyFields) {
            this.rowKeyFields = rowKeyFields;
            return this;
        }

        public Builder rowKeyFields(Field... rowKeyFields) {
            return rowKeyFields(Arrays.asList(rowKeyFields));
        }

        public Builder sortKeyFields(List<Field> sortKeyFields) {
            this.sortKeyFields = sortKeyFields;
            return this;
        }

        public Builder sortKeyFields(Field... sortKeyFields) {
            return sortKeyFields(Arrays.asList(sortKeyFields));
        }

        public Builder valueFields(List<Field> valueFields) {
            this.valueFields = valueFields;
            return this;
        }

        public Builder valueFields(Field... valueFields) {
            return valueFields(Arrays.asList(valueFields));
        }

        public Schema build() {
            return new Schema(this);
        }

        private List<Field> validRowKeyFields() {
            return validateRowKeys(rowKeyFields);
        }

        private List<Field> validSortKeyFields() {
            if (sortKeyFields == null) {
                return Collections.emptyList();
            } else {
                return validateSortKeys(sortKeyFields);
            }
        }

        private List<Field> validValueFields() {
            if (valueFields == null) {
                return Collections.emptyList();
            } else {
                return valueFields;
            }
        }
    }

    private static List<Field> validateRowKeys(List<Field> fields) {
        if (fields == null || fields.isEmpty()) {
            throw new IllegalArgumentException("Must have at least one row key field");
        }
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Row key fields must have a primitive type");
        }
        return fields;
    }

    private static List<Field> validateSortKeys(List<Field> fields) {
        if (fields.stream().anyMatch(field -> !(field.getType() instanceof PrimitiveType))) {
            throw new IllegalArgumentException("Sort key fields must have a primitive type");
        }
        return fields;
    }

    private static void validateNoDuplicates(Stream<Field> fields) {
        Set<String> foundNames = new HashSet<>();
        Set<String> duplicates = new TreeSet<>();
        fields.forEach(field -> {
            boolean isNew = foundNames.add(field.getName());
            if (!isNew) {
                duplicates.add(field.getName());
            }
        });
        if (!duplicates.isEmpty()) {
            throw new IllegalArgumentException("Found duplicate field names: " + duplicates);
        }
    }
}
