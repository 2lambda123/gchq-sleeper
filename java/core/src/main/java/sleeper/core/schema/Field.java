/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.schema;

import sleeper.core.schema.type.Type;

import java.util.Objects;

/**
 * Describes a field in the {@link Schema}. It contains a {@link String} name
 * and a {@link Type}.
 */
public class Field {
    private String name;
    private Type type;
    
    public Field() {
    }
    
    public Field(String name, Type type) {
        setName(name);
        setType(type);
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        Objects.requireNonNull(name, "name must not be null");
        this.name = name;
    }

    public Type getType() {
        return type;
    }

    public void setType(Type type) {
        Objects.requireNonNull(type, "type must not be null");
        this.type = type;
    }

    @Override
    public String toString() {
        return "Field{" + "name=" + name + ", type=" + type + '}';
    }

    @Override
    public int hashCode() {
        int hash = 5;
        hash = 31 * hash + Objects.hashCode(this.name);
        hash = 31 * hash + Objects.hashCode(this.type);
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final Field other = (Field) obj;
        if (!Objects.equals(this.name, other.name)) {
            return false;
        }
        if (!Objects.equals(this.type, other.type)) {
            return false;
        }
        return true;
    }
}
