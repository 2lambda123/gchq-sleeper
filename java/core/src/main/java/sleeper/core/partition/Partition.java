/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import sleeper.core.key.Key;
import sleeper.core.range.Region;
import sleeper.core.range.RegionCanonicaliser;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.PrimitiveType;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * A Partition is a Region of key space, with additional information that allows
 * the position of the partition in the tree of partitions to be identified,
 * e.g. it has a unique id, the id of its parent partition, the id of any child
 * partitions, and if it has been split the dimension it was split on.
 * <p>
 * The {@link Region} must be in canonical form, i.e. all the ranges must contain
 * their minimum but not contain their maximum.
 */
public class Partition {
    private List<PrimitiveType> rowKeyTypes;
    private Region region;
    private String id;
    private boolean leafPartition;
    private String parentPartitionId;
    private List<String> childPartitionIds;
    private int dimension = -1; // -1 used to indicate that it has not been split yet; when it has been split, indicates which dimension was used to split on.

    public Partition() {
    }

    public Partition(
            List<PrimitiveType> rowKeyTypes,
            Region region,
            String id,
            boolean leafPartition,
            String parentPartitionId,
            List<String> childPartitionIds,
            int dimension) {
        setRowKeyTypes(rowKeyTypes);
        setRegion(region);
        setId(id);
        setLeafPartition(leafPartition);
        setParentPartitionId(parentPartitionId);
        setChildPartitionIds(childPartitionIds);
        setDimension(dimension);
    }

    public List<PrimitiveType> getRowKeyTypes() {
        return rowKeyTypes;
    }

    public void setRowKeyTypes(List<PrimitiveType> rowKeyTypes) {
        this.rowKeyTypes = rowKeyTypes;
    }

    public void setRowKeyTypes(PrimitiveType... rowKeyTypes) {
        this.rowKeyTypes = new ArrayList<>();
        for (PrimitiveType type : rowKeyTypes) {
            this.rowKeyTypes.add(type);
        }
    }

    public Region getRegion() {
        return region;
    }

    public void setRegion(Region region) {
        if (!RegionCanonicaliser.isRegionInCanonicalForm(region)) {
            throw new IllegalArgumentException("Region must be in canonical form");
        }
        this.region = region;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public boolean isLeafPartition() {
        return leafPartition;
    }

    public void setLeafPartition(boolean leafPartition) {
        this.leafPartition = leafPartition;
    }

    public String getParentPartitionId() {
        return parentPartitionId;
    }

    public void setParentPartitionId(String parentPartitionId) {
        this.parentPartitionId = parentPartitionId;
    }

    public List<String> getChildPartitionIds() {
        if (null == childPartitionIds) {
            childPartitionIds = new ArrayList<>();
        }
        return childPartitionIds;
    }

    public void setChildPartitionIds(List<String> childPartitionIds) {
        this.childPartitionIds = new ArrayList<>();
        if (null != childPartitionIds) {
            this.childPartitionIds.addAll(childPartitionIds);
        }
    }

    public int getDimension() {
        return dimension;
    }

    public void setDimension(int dimension) {
        this.dimension = dimension;
    }

    public boolean isRowKeyInPartition(Schema schema, Key rowKey) {
        return region.isKeyInRegion(schema, rowKey);
    }

    public boolean doesRegionOverlapPartition(Region otherRegion) {
        return region.doesRegionOverlap(otherRegion);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Partition partition = (Partition) o;

        return Objects.equals(rowKeyTypes, partition.rowKeyTypes)
                && leafPartition == partition.leafPartition
                && Objects.equals(region, partition.region)
                && Objects.equals(id, partition.id)
                && Objects.equals(parentPartitionId, partition.parentPartitionId)
                && Objects.equals(getChildPartitionIds(), partition.getChildPartitionIds())
                && dimension == partition.getDimension();
    }

    @Override
    public int hashCode() {
        return Objects.hash(rowKeyTypes, region, id, leafPartition,
                parentPartitionId, getChildPartitionIds(), dimension);
    }

    @Override
    public String toString() {
        return "Partition{"
                + "rowKeyTypes=" + rowKeyTypes
                + ", region=" + region
                + ", id='" + id + '\''
                + ", leafPartition=" + leafPartition
                + ", parentPartitionId='" + parentPartitionId + '\''
                + ", childPartitionIds=" + getChildPartitionIds()
                + ", dimension=" + dimension
                + '}';
    }
}
