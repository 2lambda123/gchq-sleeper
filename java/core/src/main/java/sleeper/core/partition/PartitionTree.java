/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import sleeper.core.key.Key;
import sleeper.core.schema.Schema;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * This represents a tree of {@link Partition}s. It can be used to find all
 * ancestors of a partition, i.e. all partitions that are either parents of the
 * partition, or grandparents, or great-grandparents.
 */
public class PartitionTree {
    private final Schema schema;
    private final Map<String, Partition> idToPartition;
    private final Partition rootPartition;

    public PartitionTree(Schema schema, List<Partition> partitions) {
        this.schema = schema;
        this.idToPartition = new HashMap<>();
        partitions.forEach(p -> this.idToPartition.put(p.getId(), p));
        List<Partition> rootPartitions = partitions.stream().filter(p -> null == p.getParentPartitionId()).collect(Collectors.toList());
        // There should be exactly one root partition.
        if (rootPartitions.size() != 1) {
            throw new IllegalArgumentException("There should be exactly one root partition, found " + rootPartitions.size());
        }
        this.rootPartition = rootPartitions.get(0);
    }

    public List<String> getChildIds(String partitionId) throws IllegalArgumentException {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }
        return idToPartition.get(partitionId).getChildPartitionIds();
    }

    public String getParentId(String partitionId) {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }
        return idToPartition.get(partitionId).getParentPartitionId();
    }

    public List<String> getAllAncestorIds(String partitionId) {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }

        List<String> ancestors = new ArrayList<>();
        String parent = idToPartition.get(partitionId).getParentPartitionId();
        while (null != parent) {
            ancestors.add(parent);
            parent = null == idToPartition.get(parent) ? null : idToPartition.get(parent).getParentPartitionId();
        }
        return ancestors;
    }

    public List<Partition> getAllAncestors(String partitionId) {
        if (!idToPartition.containsKey(partitionId)) {
            throw new IllegalArgumentException("No partition of id " + partitionId);
        }

        List<Partition> ancestors = new ArrayList<>();
        String parentId = idToPartition.get(partitionId).getParentPartitionId();
        while (null != parentId) {
            ancestors.add(idToPartition.get(parentId));
            parentId = idToPartition.get(parentId).getParentPartitionId();
        }
        return ancestors;
    }

    public Partition getPartition(String partitionId) {
        return idToPartition.get(partitionId);
    }

    public List<Partition> getAllPartitions() {
        return Collections.unmodifiableList(new ArrayList<>(idToPartition.values()));
    }

    public List<Partition> getLeafPartitions() {
        return Collections.unmodifiableList(idToPartition.values().stream()
                .filter(Partition::isLeafPartition)
                .collect(Collectors.toList()));
    }

    public Partition getLeafPartition(Key key) {
        // Sanity check key is of the correct length
        if (key.size() != schema.getRowKeyFields().size()) {
            throw new IllegalArgumentException("Key must match the row key fields from the schema (key was "
                    + key + ", schema has row key fields " + schema.getRowKeyFields() + ")");
        }
        // If root partition is a leaf partition then key must be in it
        if (rootPartition.isLeafPartition()) {
            return rootPartition;
        }

        return descend(rootPartition, key);
    }

    private Partition descend(Partition currentNode, Key key) {
        // Get child partitions
        List<String> childPartitionIds = currentNode.getChildPartitionIds();
        List<Partition> childPartitions = new ArrayList<>();
        childPartitionIds.forEach(c -> childPartitions.add(idToPartition.get(c)));

        // Which child is the key in?
        Partition child = null;
        for (Partition partition : childPartitions) {
            if (partition.isRowKeyInPartition(schema, key)) {
                child = partition;
                break;
            }
        }
        if (null == child) {
            throw new IllegalArgumentException("Found key that was not in any of the child partitions: key " + key
                    + ", child partitions " + childPartitions);
        }
        if (child.isLeafPartition()) {
            return child;
        }
        return descend(child, key);
    }

    public Partition getRootPartition() {
        return rootPartition;
    }

    public Partition getNearestCommonAncestor(Key a, Key b) {
        return getNearestCommonAncestor(getLeafPartition(a), getLeafPartition(b));
    }

    public Partition getNearestCommonAncestor(Partition a, Partition b) {
        if (a.getId().equals(b.getId())) {
            return a;
        }
        Set<String> ancestorsB = new HashSet<>(getAllAncestorIds(b.getId()));
        for (String ancestorA : getAllAncestorIds(a.getId())) {
            if (ancestorsB.contains(ancestorA)) {
                return getPartition(ancestorA);
            }
        }
        return getRootPartition();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        PartitionTree that = (PartitionTree) o;
        return schema.equals(that.schema)
                && idToPartition.equals(that.idToPartition);
    }

    @Override
    public int hashCode() {
        return Objects.hash(schema, idToPartition);
    }
}
