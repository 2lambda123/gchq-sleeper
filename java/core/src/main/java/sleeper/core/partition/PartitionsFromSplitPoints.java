/*
 * Copyright 2022 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.core.partition;

import com.facebook.collections.ByteArray;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import sleeper.core.range.Range;
import sleeper.core.range.Range.RangeFactory;
import sleeper.core.range.Region;
import sleeper.core.schema.Field;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.ByteArrayType;
import sleeper.core.schema.type.IntType;
import sleeper.core.schema.type.LongType;
import sleeper.core.schema.type.PrimitiveType;
import sleeper.core.schema.type.StringType;
import sleeper.core.schema.type.Type;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

/**
 * Given a list of split points that split the first dimension of the row keys into partitions, this class
 * constructs a full tree of partitions, up to a single root.
 */
public class PartitionsFromSplitPoints {
    private static final Logger LOGGER = LoggerFactory.getLogger(PartitionsFromSplitPoints.class);

    private final Schema schema;
    private final List<Field> rowKeyFields;
    private final List<PrimitiveType> rowKeyTypes;
    private final List<Object> splitPoints;
    private final RangeFactory rangeFactory;

    public PartitionsFromSplitPoints(
            Schema schema, List<Object> splitPoints) {
        this.schema = schema;
        this.rowKeyFields = schema.getRowKeyFields();
        this.rowKeyTypes = new ArrayList<>();
        for (Field field : rowKeyFields) {
            this.rowKeyTypes.add((PrimitiveType) field.getType());
        }
        this.splitPoints = splitPoints;
        this.rangeFactory = new RangeFactory(schema);
    }

    public List<Partition> construct() {
        // If there are no split points then create a single root partition, which covers the entire key space, and
        // is a leaf partition.
        if (null == splitPoints || splitPoints.isEmpty()) {
            LOGGER.info("Constructing partition tree from no split points - tree will consist of one partition");
            return Collections.singletonList(createRootPartitionThatIsLeaf());
        }

        validateSplitPoints();
        LOGGER.info("Split points are valid");

        // There is at least 1 split point. Use the split points to create leaf partitions.
        List<Partition> leafPartitions = createLeafPartitions();
        List<Partition> allPartitions = new ArrayList<>(leafPartitions);

        List<Partition> nextLayer = addLayer(leafPartitions, allPartitions);
        while (1 != nextLayer.size()) {
            nextLayer = addLayer(nextLayer, allPartitions);
        }

        return allPartitions;
    }

    private List<Partition> addLayer(List<Partition> partitionsInLayer, List<Partition> allPartitions) {
        List<Partition> parents = new ArrayList<>();
        for (int i = 0; i < partitionsInLayer.size(); i += 2) {
            if (i <= partitionsInLayer.size() - 2) {
                Partition leftPartition = partitionsInLayer.get(i);
                Partition rightPartition = partitionsInLayer.get(i + 1);

                Partition parent = new Partition();
                parent.setId(UUID.randomUUID().toString());
                parent.setParentPartitionId(null);
                parent.setChildPartitionIds(Arrays.asList(leftPartition.getId(), rightPartition.getId()));
                parent.setLeafPartition(false);
                parent.setDimension(0);
                parent.setRowKeyTypes(leftPartition.getRowKeyTypes());
                List<Range> ranges = new ArrayList<>();
                for (Range range : leftPartition.getRegion().getRanges()) {
                    if (!range.getFieldName().equals(rowKeyFields.get(0).getName())) {
                        ranges.add(range); // TODO Check that left and right have the same ranges in the dimensions other than 0
                    }
                }
                Range rangeForDim0 = rangeFactory.createRange(rowKeyFields.get(0),
                        leftPartition.getRegion().getRange(rowKeyFields.get(0).getName()).getMin(),
                        true,
                        rightPartition.getRegion().getRange(rowKeyFields.get(0).getName()).getMax(),
                        false);
                ranges.add(rangeForDim0);
                Region region = new Region(ranges);
                parent.setRegion(region);

                leftPartition.setParentPartitionId(parent.getId());
                rightPartition.setParentPartitionId(parent.getId());

                parents.add(parent);
            }
        }
        allPartitions.addAll(parents);

        // If there were an odd number of partitions in partitionsInLayer then we need to add the remaining one into
        // the next layer, but it shouldn't be added into allPartitions as it is already there.
        if (partitionsInLayer.size() % 2 == 1) {
            parents.add(partitionsInLayer.get(partitionsInLayer.size() - 1));
        }

        LOGGER.info("Created layer of {} partitions from previous layer of {} partitions", parents.size(), partitionsInLayer.size());
        LOGGER.debug("New partitions are {}", parents);

        return parents;
    }

    private List<Partition> createLeafPartitions() {
        List<Region> leafRegions = leafRegionsFromSplitPoints(schema, splitPoints);
        List<Partition> leafPartitions = new ArrayList<>();
        for (Region region : leafRegions) {
            Partition partition = new Partition();
            partition.setRowKeyTypes(rowKeyTypes);
            partition.setRegion(region);
            partition.setId(UUID.randomUUID().toString());
            partition.setLeafPartition(true);
            partition.setParentPartitionId(null);
            partition.setChildPartitionIds(new ArrayList<>());
            partition.setDimension(-1);
            leafPartitions.add(partition);
        }
        LOGGER.info("Created {} leaf partitions from {} split points", leafPartitions.size(), splitPoints.size());
        LOGGER.debug("Partitions are {}", leafPartitions);
        return leafPartitions;
    }

    private Partition createRootPartitionThatIsLeaf() {
        Partition rootPartition = new Partition();
        rootPartition.setRowKeyTypes(rowKeyTypes);
        List<Range> ranges = new ArrayList<>();
        for (Field field : rowKeyFields) {
            ranges.add(getRangeCoveringWholeDimension(field));
        }
        Region region = new Region(ranges);
        rootPartition.setRegion(region);
        rootPartition.setId("root");
        rootPartition.setLeafPartition(true);
        rootPartition.setParentPartitionId(null);
        rootPartition.setChildPartitionIds(new ArrayList<>());
        rootPartition.setDimension(-1);
        return rootPartition;
    }

    private Range getRangeCoveringWholeDimension(Field field) {
        Range range = rangeFactory.createRange(field, getMinimum(field.getType()), true, null, false);
        return range;
    }

    private static Object getMinimum(Type type) {
        if (type instanceof IntType) {
            return Integer.MIN_VALUE;
        }
        if (type instanceof LongType) {
            return Long.MIN_VALUE;
        }
        if (type instanceof StringType) {
            return "";
        }
        if (type instanceof ByteArrayType) {
            return new byte[]{};
        }
        throw new IllegalArgumentException("Unknown key type " + type);
    }

    private void validateSplitPoints() {
        int count = 0;
        Comparable previous = null;
        for (Object obj : splitPoints) {
            validateCorrectType(obj);
            Comparable comparable = getAsComparable(obj);
            if (count > 0) {
                if (previous.compareTo(comparable) >= 0) {
                    throw new IllegalArgumentException("Invalid split point: " + previous + " should be less than " + comparable);
                }
            }
            previous = comparable;
            count++;
        }
    }

    private void validateCorrectType(Object obj) {
        Type type = rowKeyTypes.get(0);
        if (type instanceof IntType) {
            if (!(obj instanceof Integer)) {
                throw new IllegalArgumentException("Invalid split point: " + obj + " should be of type Integer");
            }
        } else if (type instanceof LongType) {
            if (!(obj instanceof Long)) {
                throw new IllegalArgumentException("Invalid split point: " + obj + " should be of type Long");
            }
        } else if (type instanceof StringType) {
            if (!(obj instanceof String)) {
                throw new IllegalArgumentException("Invalid split point: " + obj + " should be of type String");
            }
        } else if (type instanceof ByteArrayType) {
            if (!(obj instanceof byte[])) {
                throw new IllegalArgumentException("Invalid split point: " + obj + " should be of type byte[]");
            }
        } else {
            throw new IllegalArgumentException("Unknown key type " + type);
        }
    }

    private Comparable getAsComparable(Object obj) {
        Type type = rowKeyTypes.get(0);
        if (type instanceof ByteArrayType) {
            return ByteArray.wrap((byte[]) obj);
        }
        return (Comparable) obj;
    }

    public static List<Region> leafRegionsFromSplitPoints(Schema schema, List<Object> splitPoints) {
        RangeFactory rangeFactoryStatic = new RangeFactory(schema);
        List<PrimitiveType> rowKeyTypes = schema.getRowKeyTypes();
        List<Field> rowKeyFields = schema.getRowKeyFields();
        List<Object> partitionBoundaries = new ArrayList<>();
        Type type = rowKeyTypes.get(0);
        partitionBoundaries.add(getMinimum(type));
        partitionBoundaries.addAll(splitPoints);
        partitionBoundaries.add(null);

        // Create ranges for the other dimensions
        List<Range> ranges = new ArrayList<>();
        for (Field rowKeyField : rowKeyFields.subList(1, rowKeyFields.size())) {
            Type rowKeyType = rowKeyField.getType();
            Range range = rangeFactoryStatic.createRange(rowKeyField, getMinimum(rowKeyType), true, null, false);
            ranges.add(range);
        }

        List<Region> leafRegions = new ArrayList<>();
        for (int i = 0; i < partitionBoundaries.size() - 1; i++) {
            List<Range> rangesForThisRegion = new ArrayList<>();
            Range rangeForDim0 = rangeFactoryStatic.createRange(rowKeyFields.get(0), partitionBoundaries.get(i), true, partitionBoundaries.get(i + 1), false);
            rangesForThisRegion.add(rangeForDim0);
            rangesForThisRegion.addAll(ranges);
            Region region = new Region(rangesForThisRegion);
            leafRegions.add(region);
        }
        return leafRegions;
    }

    public static PartitionTree treeFrom(Schema schema, List<Object> splitPoints) {
        return new PartitionTree(schema, new PartitionsFromSplitPoints(schema, splitPoints).construct());
    }

}
