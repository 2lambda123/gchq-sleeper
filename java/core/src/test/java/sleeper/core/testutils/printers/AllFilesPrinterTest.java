
package sleeper.core.testutils.printers;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import sleeper.core.partition.PartitionsBuilder;
import sleeper.core.schema.Schema;
import sleeper.core.schema.type.StringType;
import sleeper.core.statestore.FileReferenceFactory;

import java.time.Instant;
import java.util.List;

import static sleeper.core.schema.SchemaTestHelper.schemaWithKey;
import static sleeper.core.statestore.FilesReportTestHelper.activeAndReadyForGCFilesReport;

public class AllFilesPrinterTest {

    private final Schema schema = schemaWithKey("key", new StringType());
    private final PartitionsBuilder partitions = new PartitionsBuilder(schema);

    @Test
    void shouldPrintReferencedAndUnreferencedFiles() {
        // Given
        partitions.rootFirst("root")
                .splitToNewChildren("root", "L", "R", "row-50");

        // When
        FileReferenceFactory fileReferenceFactory = fileReferenceFactory();
        String printed = AllFilesPrinter.printFiles(partitions.buildTree(),
                activeAndReadyForGCFilesReport(Instant.now(),
                        List.of(fileReferenceFactory.partitionFile("L", 10)),
                        List.of("oldFile1.parquet", "oldFile2.parquet")));

        // Then see approved output
        Approvals.verify(printed);
    }

    private FileReferenceFactory fileReferenceFactory() {
        return FileReferenceFactory.from(partitions.buildTree());
    }
}
