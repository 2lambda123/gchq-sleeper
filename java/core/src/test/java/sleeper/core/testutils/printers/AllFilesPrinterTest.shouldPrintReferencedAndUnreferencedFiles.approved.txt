Unreferenced files: 2

Active files:
Partition at L: 10 records in file
